robot/jobmanager/jobmanager.go

```diff
@@ -48,6 +52,61 @@ type JobManager struct {
 	conn          rpc.ClientConn
 	isClosed      bool
 	closeMutex    sync.Mutex
+
+	NumJobHistories atomic.Int32
+	JobHistories    ssync.Map[string, *JobHistory]
+}
+
+// JobHistory records historical metadata about a job.
+type JobHistory struct {
+	successTimesMu sync.Mutex
+	successTimes   *ring.Ring
+	failureTimesMu sync.Mutex
+	failureTimes   *ring.Ring
+}
+
+// Successes returns timestamps of the last historyLength number successfully completed jobs.
+func (jh *JobHistory) Successes() []time.Time {
+	ts := make([]time.Time, 0, historyLength)
+	jh.successTimesMu.Lock()
+	defer jh.successTimesMu.Unlock()
+	for i := 0; i < historyLength; i++ {
+		if jh.successTimes.Value != nil {
+			ts = append(ts, jh.successTimes.Value.(time.Time))
+		}
+		jh.successTimes = jh.successTimes.Next()
```

Comment:
I think I'm a little lost on how a `Ring` works, or more specifically how your `AddFoo` and `Foo` functions are working. `AddSuccess` seems to set the current value and then point the value to `Next()`, but this `Successes` method seem to get the "last historyLength number jobs" by getting the current success and then using `Next()` again. Wouldn't that `Next()` start us over at the beginning of the ring?